resources/kernel/make_kernel.sh

```diff
@@ -0,0 +1,185 @@
+    [ ! -d "$busybox_rootfs" ] && die "Busybox rootfs directory not found."
+
+    # Move to the directory with the kernel sources.
+    pushd_quiet "$kernel_dir"
+
+    # Prepare initramfs directory.
+    mkdir -p initramfs/{bin,dev,etc,home,mnt,proc,sys,usr}
+    # Copy busybox.
+    echo "Copying busybox to the initramfs directory..."
+    cp -r "$busybox_rootfs"/* initramfs/
+
+    # Make a block device and a console.
+    pushd_quiet initramfs/dev
+    echo "Creating device nodes..."
+    rm -f sda && mknod sda b 8 0
+    rm -f console && mknod console c 5 1
+    rm -f ttyS0 && mknod ttyS0 c 4 64
+
+    make_init "$halt"
+
+    chmod +x init
+    fakeroot chown root init
+
+    # Pack it up...
+    echo "Packing initramfs.cpio..."
+    find . | cpio -H newc -o > ../initramfs.cpio
+    fakeroot chown root ../initramfs.cpio
+
+    # Return to kernel srcdir.
+    popd_quiet
+    # Return to previous directory.
+    popd_quiet
+}
+
+# Usage: validate_kernel_format format
+# Prints the lowercase format name, if one of "elf" or "bzimage".
+# Exits with error if any other format is specified.
+validate_kernel_format() {
+    format="$1"
+
+    kernel_fmt=$(echo "$format" | tr '[:upper:]' '[:lower:]')
+    if [ "$kernel_fmt" != "elf" ] && [ "$kernel_fmt" != "bzimage" ]; then
+        die "Invalid kernel binary format: $kernel_fmt."
+    fi
+    echo "$kernel_fmt"
+}
+
+# Usage: kernel_target format
+# Prints the `make` target that builds a kernel of the specified format.
+kernel_target() {
+    format=$(validate_kernel_format "$1")
+
+    case "$format" in
+    elf)        echo "vmlinux"
+                ;;
+    bzimage)    # This is the default target.
+                ;;
+    esac
+}
+
+# Usage: kernel_binary format
+# Prints the name of the generated kernel binary.
+kernel_binary() {
+    format=$(validate_kernel_format "$1")
+    
+    case "$format" in
+    elf)        echo "vmlinux"
+                ;;
+    bzimage)    echo "arch/x86/boot/bzImage"
+                ;;
+    esac
+}
+
+# Usage:
+#   make_kernel
+#       /path/to/kernel/dir \
+#       format              \
+#       [num_cpus_build]    \
+#       [/path/to/kernel/destination]
+make_kernel() {
+    kernel_dir="$1"
+    format="$2"
+    nprocs="$3"
+    dst="$4"
+
+    [ -z "$kernel_dir" ] && die "Kernel directory not specified."
+    [ ! -d "$kernel_dir" ] && die "Kernel directory not found."
+    [ -z "$format" ] && die "Kernel format not specified."
+    [ -z "$nprocs" ] && nprocs=1
+   
+    target=$(kernel_target "$format")
+    kernel_binary=$(kernel_binary "$format")
+
+    # Move to the directory with the kernel sources.
+    pushd_quiet "$kernel_dir"
+
+    # Build kernel.
+    echo "Building kernel..."
+    # Missing the quotes for $target is intentional: if the target is empty,
+    # quotes will cause it to be passed as an empty string, which `make`
+    # doesn't understand. No quotes => no empty string.
+    make -j "$nprocs" $target
+
+    if [ -n "$dst" ]; then
```

Comment:
We should do the `cp` only if "$kernel_binary" and "$dst" are also different, otherwise if "$dst" is "vmlinux" for example, the `cp` will give the following error: `cp: 'vmlinux' and 'vmlinux' are the same file` (this happens if the user passes `vmlinux` as value to `-k` or doesn't provide `-k` at all).

Reply:
Also, it looks like the way we are using make_kernel(), `"$dst"` is always present. Would it make sense to remove [this line](https://github.com/rust-vmm/vmm-reference/pull/21/commits/793e04ea5f0e438a8f20593402e3bc4c0fdf2569#diff-26cd4e8b4eaa05f96c140feb26fb78afe7f0c4c21b8cfc526c4b90013459c231R86)? This way we would `cp` only if user provided a kernel image name.

Reply:
`KERNEL_BINARY` has to be *something* for the line immediately after that, which appends the `-halt` suffix if the user wants a halt-y kernel. I'll rework the `cp`, good catch on the error!

Reply:
Hmm, can we add a mention about the suffix appending to [this description](https://github.com/rust-vmm/vmm-reference/pull/21/files#diff-26cd4e8b4eaa05f96c140feb26fb78afe7f0c4c21b8cfc526c4b90013459c231R47)? I would expect the resulting kernel image to have the exact name that I am passing to `-k` option.